### ISSUE TITLE
How to list offline members that have a specific role ?

### ISSUE THREAD

**Issue by captain-majid** [Merubokkusu/Discord-S.C.U.M#148]
Hi there, I'm trying to list all users with "mod" or "admin" roles, but bc the guild is > 1000 members, I can't see offline roles (mods/admins), any solution for that ?
```python
import discum

bot = discum.Client(token='myToken')
guild_id = '426172214677602304'
channel_id = '426172214677602306' 

bot.gateway.fetchMembers(guild_id, channel_id, keep=['username', 'roles'], method='overlap')
@bot.gateway.command
def memberTest(resp):
    if bot.gateway.finishedMemberFetching(guild_id):
        bot.gateway.removeCommand(memberTest)
        bot.gateway.close()
bot.gateway.run()

for memberID in bot.gateway.session.guild(guild_id).members:
    userName    = bot.gateway.session.guild(guild_id).members[memberID]['username']
    memberRoles = bot.gateway.session.guild(guild_id).members[memberID]['roles']
    role1 = '656609109365358612' # mod
    role2 = '296936984994119681' # admin
    allRoles = bot.gateway.session.guild(guild_id).roles
    if (role2 in memberRoles):
        print(userName + ' [' + allRoles[role2]['name'] + ']') 
    if (role1 in memberRoles):
        print(userName + ' [' + allRoles[role1]['name'] + ']') 
```
_____________
arandomnewaccount:
according to https://github.com/Merubokkusu/Discord-S.C.U.M/blob/master/docs/using/fetchingGuildMembers.md bot.gateway.fetchMembers only gets the online users (for large servers). To get offline users as well, use op8 brute forcing (here's an example: https://github.com/Merubokkusu/Discord-S.C.U.M/blob/master/docs/using/fetchingGuildMembers.md#search-for-members-by-query)

**Comment by captain-majid** [comment]
Sorry, I tried to play with the code there, but had no luck!
arandomnewaccount:
to fetch members, copy lines 31-136 from https://github.com/Merubokkusu/Discord-S.C.U.M/blob/master/examples/searchGuildMembers.py

Then set the guildID variable and wait variable. Wait is just wait time inbetween responses and msgs.
```
guildID = "guild ID here"
wait=1
```

Then, run the brute forcer:
```
bot.gateway.command({"function": bruteForceTest, "params":{"guildID":guildID, "wait":wait}})
bot.gateway.run()
```

Now, wait a long time for the brute forcer to finish.

Once it's finished, parse the members like you usually do:
```python
for memberID in bot.gateway.session.guild(guild_id).members:
    ...
```

**Comment by cth-latest** [comment]
If i'm not wrong it's not possible as discord don't send the offlines users if the server has more than 1000 members.
____________
arandomnewaccount:
I wrote an answer. You can check it out. It is possible and the fastest way to do it (that I know of) is opcode 8 brute forcing. 
Only for opcode 14 discord doesn't send offline users.

**Comment by cth-latest** [comment]
I think it's really really suspicious and can up the chances to get locked. The client never request the members like that. (Except for owners)
_________________________
arandomnewaccount:
"The client can never request the members like that"
This is incorrect. Whenever you search for members (either in the search bar of a guild or in the home screen), your client uses op 8 with a query and a nonzero limit. 

Owners of guilds also use op8 to get members, but instead of using a query and a nonzero limit, they use a query="" and a limit=0.

As for chances of getting locked, idk. I've tested the brute forcer multiple times (on really large guilds aka >100k members) without problems. 

edit: I think there's an endpoint to get the userIDs that have a role...Ill add that in soon.

**Comment by captain-majid** [comment]
> Now, wait a long time for the brute forcer to finish.
> 
> Once it's finished, parse the members like you usually do:
> 
> ```python
> for memberID in bot.gateway.session.guild(guild_id).members:
>     ...
> ```
I did this code, but I just see the cmd window logging some random gateway events slowly, I'm doing this with a small guild (~1500 members), but there is nothing to say when the code is done, any indicator or amount of time expected there for the process to finish ?, or to show that I'm doing ok ?:
```py
#EXAMPLE 3: opcode 8 brute forcer
#not entirely random. Optimized quite a bit.

import time
import re
import discum
import Discord_tests

bot = discum.Client(token=Discord_tests.myToken)
allchars = [' ', '!', '"', '#', '$', '%', '&', "'", '(', ')', '*', '+', ',', '-', '.', '/', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', ':', ';', '<', '=', '>', '?', '@', '[', ']', '^', '_', '`', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', '{', '|', '}', '~']
bot.qList = ["!"] #query list
bot.gateway.guildMemberSearches = {}
bot.gateway.resetMembersOnSessionReconnect = False #member list brute forcing can take a while

guildID = '426172214677602304' # CPL
channel_id = '426172214677602306' 

class MemberFetchingScore:
	def __init__(self, perRequestExpectation, perSecondExpectation):
		self.perRequestExpectation = perRequestExpectation #expected # of members per request
		self.perSecondExpectation = perSecondExpectation #expected # of members per second
		self.effectiveness = 0
		self.efficiency = 0
		self.completeness = 0
	#percentage of requests returning back expected # of members
	def calculateEffectiveness(self, guildID):
		self.effectiveness = 100*(len(bot.gateway.session.guild(guildID).members)/len(bot.gateway.guildMemberSearches[guildID]["queries"]))/self.perRequestExpectation
	#percentage of expected members fetched per second
	def calculateEfficiency(self, guildID, startTime):
		totalTime = time.time() - startTime
		self.efficiency = 100*(len(bot.gateway.session.guild(guildID).members)/totalTime)/self.perSecondExpectation
	#percentage of members fetched over total members in server
	def calculateCompleteness(self, guildID):
		self.completeness = 100*(len(bot.gateway.session.guild(guildID).members)/bot.gateway.session.guild(guildID).memberCount)
	#average of measures, 0<=score<=100
	def getScore(self):
		return (self.effectiveness+self.efficiency+self.completeness)/3

s = MemberFetchingScore(100, 100)

def calculateOption(guildID, action): #action == 'append' or 'replace'
	if action == 'append':
		lastUserIDs = bot.gateway.guildMemberSearches[guildID]["queries"][''.join(bot.qList)]
		data = [bot.gateway.session.guild(guildID).members[i] for i in bot.gateway.session.guild(guildID).members if i in lastUserIDs]
		lastName = sorted(set([re.sub(' +', ' ', j['nick'].lower()) if (j.get('nick') and re.sub(' +', ' ', j.get('nick').lower()).startswith(''.join(bot.qList))) else re.sub(' +', ' ', j['username'].lower()) for j in data]))[-1]
		try:
			option = lastName[len(bot.qList)]
			return option
		except IndexError:
			return None
	elif action == 'replace':
		if bot.qList[-1] in allchars:
			options = allchars[allchars.index(bot.qList[-1])+1:]
			if ' ' in options and (len(bot.qList)==1 or (len(bot.qList)>1 and bot.qList[-2]==' ')): #cannot start with a space and cannot have duplicate spaces
				options.remove(' ')
			return options
		else:
			return None

def findReplaceableIndex(guildID):
	for i in range(len(bot.qList)-2, -1, -1): #assume that the last index is not changable
		if bot.qList[i] != '~':
			return i
	return None

def bruteForceTest(resp, guildID, wait):
	if resp.event.ready_supplemental:
		bot.startTime = time.time()
		bot.gateway.queryGuildMembers([guildID], query=''.join(bot.qList), limit=100, keep="all")
	elif resp.event.guild_members_chunk:
		remove = False
		if len(bot.gateway.guildMemberSearches[guildID]["queries"][''.join(bot.qList)]) == 100: #append
			appendOption = calculateOption(guildID, 'append')
			if appendOption:
				bot.qList.append(appendOption)
			else:
				remove = True
		else: #if <100 results returned, replace
			replaceOptions = calculateOption(guildID, 'replace')
			if replaceOptions:
				bot.qList[-1] = replaceOptions[0]
			else:
				remove = True
		if remove: #if no replace options, find first replaceable index & replace it
			if len(bot.qList) == 1: #reached end of possibilities
				bot.gateway.removeCommand({"function": bruteForceTest, "params":{"guildID":guildID, "wait":wait}})
				s.calculateEfficiency(guildID, bot.startTime)
				print("efficiency: "+repr(s.efficiency)+"%")
				s.calculateCompleteness(guildID)
				print("completeness: "+repr(s.completeness)+"%")
				print("score: "+repr(s.getScore()))
				bot.gateway.close()
			else:
				replaceableInd = findReplaceableIndex(guildID)
				if replaceableInd != None:
					bot.qList = bot.qList[:replaceableInd+1]
					replaceOptions = calculateOption(guildID, 'replace')
					bot.qList[-1] = replaceOptions[0]
				else:
					bot.gateway.removeCommand({"function": bruteForceTest, "params":{"guildID":guildID, "wait":wait}})
					s.calculateEfficiency(guildID, bot.startTime)
					print("efficiency: "+repr(s.efficiency)+"%")
					s.calculateCompleteness(guildID)
					print("completeness: "+repr(s.completeness)+"%")
					print("score: "+repr(s.getScore()))
					bot.gateway.close()
		if wait: time.sleep(wait)
		print("next query: "+"".join(bot.qList))
		print("members fetched so far: "+repr(len(bot.gateway.session.guild(guildID).members)))
		s.calculateEffectiveness(guildID)
		print("effectiveness: "+repr(s.effectiveness)+"%")
		bot.gateway.queryGuildMembers([guildID], query=''.join(bot.qList), limit=100, keep="all")

wait = 1
bot.gateway.command({"function": bruteForceTest, "params":{"guildID":guildID, "wait":wait}})
bot.gateway.run()

for memberID in bot.gateway.session.guild(guildID).members:
	print(memberID)
```
_______________________________
arandomnewaccount:
It'll eventually finish and automatically close the gateway. Because it's a query brute forcer, its speed depends not only on the # of members, but also on the # of letters in the character bank (the allchars variable at the top).

It's not the best op8 brute forcing algorithm, so feel free to improve it and make it faster. You can set wait time to a lower number (I'd recommend anything above 0.5 seconds). 

Here's a brief explanation of the algorithm: https://github.com/Merubokkusu/Discord-S.C.U.M/blob/master/docs/using/fetchingGuildMembers.md#algorithm-1

edit: alr, update: in the development branch there's a new api wrap:
```python
memberIDs = bot.getRoleMemberIDs('guildID', 'roleID').json()
```
This should return all the member IDs associated with that role. Note, this does not work for the ```@everyone``` role.

Then, to check if the member if offline, you can use [bot.gateway.checkGuildMembers](https://github.com/Merubokkusu/Discord-S.C.U.M/blob/development/docs/using/Gateway_Actions.md#gatewaycheckguildmembers)

**Comment by captain-majid** [comment]
Cool new features, thanks.

When I tried with another new token (that just happened by luck) the op8 brute-forcing worked correctly/expectedly, the other token that I was testing with still doesn't work, you won't believe why is that !!!

I didn't join that specific guild on the old token, but on the new token I did !!!, I can't believe how stupid am I, thank you so much.
______________
arandomnewaccount:
lol no worries, been there done that

let me know if this was helpful so I can close the issue.
